backends/ebpf/psa/ebpfPsaGen.cpp

```diff
@@ -690,6 +690,107 @@ bool ConvertToEBPFControlPSA::preorder(const IR::ExternBlock* instance) {
+    cstring registerName = cstring::empty;
+
+    if (ann) {
+        int registerReadCounter = 0;
+        int registerWriteCounter = 0;
+        cstring indexName = cstring::empty;
+        for (auto comp : b->components) {
+            const IR::MethodCallExpression *mExpr = nullptr;
+            if (auto assign = comp->to<IR::AssignmentStatement>()) {
+                mExpr = assign->right->to<IR::MethodCallExpression>();
+            } else if (auto stat = comp->to<IR::MethodCallStatement>()) {
+                mExpr = stat->methodCall;
+            }
+
+            if (mExpr) {
+                auto mi = P4::MethodInstance::resolve(mExpr,
+                                                      control->program->refMap,
+                                                      control->program->typeMap);
+                auto ext = mi->to<P4::ExternMethod>();
+                if (ext != nullptr) {
+                    auto decl = ext->object;
+                    auto declType = ext->originalExternType;
+                    if (declType->name.name == "Hash")
+                        continue;
+                    else if (declType->name.name == "Register") {
+                        if (registerName.isNullOrEmpty())
+                            registerName = EBPFObject::externalName(decl);
+
+                        auto indexArgExpr = ext->expr->arguments->at(0)->
+                                expression->to<IR::PathExpression>();
+                        if (registerName != EBPFObject::externalName(decl)) {
+                            ::error(ErrorType::ERR_UNSUPPORTED_ON_TARGET,
+                                    "%1%: You can handle only one register "
+                                    "in the same atomic block", mExpr);
+                            return;
+                        } else if (ext->method->name.name == "read") {
+                            indexName = indexArgExpr->path->name.name;
+                            if (++registerReadCounter > 1) {
+                                ::error(ErrorType::ERR_UNSUPPORTED_ON_TARGET,
+                                        "%1%: You can handle only one register read operation"
+                                        " in the same atomic block", mExpr);
+                                return;
+                            }
+                        } else if (ext->method->name.name == "write") {
+                            if (++registerWriteCounter > 1 && registerReadCounter > 0) {
+                                ::error(ErrorType::ERR_UNSUPPORTED_ON_TARGET,
+                                        "%1%: You can handle only one register write operation"
+                                        " after register read "
+                                        "in the same atomic block", mExpr);
+                                return;
+                            }
+                            if (indexName != indexArgExpr->path->name.name) {
```

Comment:
the fact that you use the same expression does not guarantee that its value hasn't changed.
consider the example: `reg.read(i); i = i+1; reg.write(i, 0);`

Reply:
I've added more checks.